fastestimator/trace/io/csv_logger.py

```diff
@@ -12,66 +12,45 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ==============================================================================
-from typing import Optional, List
-
 import os
+from typing import List, Set, Union
+
+import pandas as pd
 
 from fastestimator.trace import Trace
-from fastestimator.util.util import is_number, to_list
 from fastestimator.util import Data
+from fastestimator.util.util import to_list
 
 
 class CSVLogger(Trace):
-    """Log monitored quantity in CSV file manner
+    """Log monitored quantity in CSV file
     Args:
         filename: Output filename.
-        monitor_names: List of key names to monitor. The names can be
-            {"mode", "epoch", "train_step", or output names that other traces
-            create}. If None, it will record all. Defaults to None.
-        separator: Seperator for numbers. Defaults to ", ".
-        append: If true, it will write csv file in append mode. Otherwise, it
-            will overwrite the existed file. Defaults to False.
-        mode: Restrict the trace to run only on given modes {'train', 'eval',
-            'test'}. None will always execute. Defaults to 'eval'.
+        monitor_names: List of key names to monitor.
+        mode: Restrict the trace to run only on given modes. None will always execute.
     """
     def __init__(self,
                  filename: str,
-                 monitor_names: Optional[List[str]] = None,
-                 separator: str = ", ",
-                 append: bool = False,
-                 mode: str = "eval"):
-        self.keys = monitor_names if monitor_names is None else to_list(monitor_names)
-        super().__init__(inputs="*" if self.keys is None else monitor_names, mode=mode)
-        self.separator = separator
+                 monitor_names: Union[List[str], str],
+                 mode: Union[str, Set[str]] = ("eval", "test")):
+        super().__init__(inputs=to_list(monitor_names), mode=mode)
         self.filename = filename
-        self.append = append
+        self.data = None
 
     def on_begin(self, data: Data):
-        self.file = open(self.filename, 'a' if self.append else 'w')
-        self.file_empty = os.stat(self.filename).st_size == 0
+        self.data = {key: [] for key in self.inputs}
```

Comment:
make self.data a defaultdict, then don't need to do this initialization

Reply:
what do you mean

Reply:
instead of self.data = {key: [].... } you can just say self.data = defaultdict(list) in the init method. then just always invoke .append() later

Reply:
done